# **Introduction** 

## **How to use?**

You can load any RAG system from our module (`rag`) where the pipeline is already implemented. The documents from the 5 TSFs have already been vectorized and indexed to the database, so you just need to indicate the path of these vectors.

The RAG systems implemented are:

1. **Naive RAG**
2. **Advanced RAG**
3. **Hybrid RAG**
4. **Advanced-Hybrid RAG**
5. **AsyncMultiQuery RAG**

In [1]:
from rag import NaiveRAG

c:\Users\gol_m\anaconda3\envs\paper_geotecnia\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


resource module not available on Windows


In [3]:
path_vector_storage = "./src/evaluation_retrieval/vector_storage/03_cenizas_espesado/03_cenizas_nomic"
embedding_type = "nomic"
embedding_model_name = "nomic-embed-text-v1.5"

pipeline_rag = NaiveRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name
)

In [ ]:
response = pipeline_rag.query("Tipo de depósito?")
response

Response(response='El tipo de depósito mencionado en el contexto es un "Depósito de Relaves Espesados". Este depósito tiene una superficie aproximada de 50 hectáreas y una pendiente media de 3%. La capacidad de depositación es de 7.600.000 m³, lo que equivale a 12,4 millones de toneladas de relaves espesados. \n\nEl diseño del depósito incluye componentes como un muro de confinamiento, un vertedero de descarga, una piscina de aguas claras, una planta de espesamiento y floculación, así como tuberías de descarga de relaves y de recuperación de aguas. La tasa de depositación de relaves espesados es de 2.929 toneladas por día.\n\nAdemás, se considera que el relave será depositado en capas homogéneas, guiado mediante camellones de tierra, y se proyecta una pendiente de depositación de 3%, aunque la pendiente natural de depositación del relave es de 7%. \n\nSi necesitas más información específica sobre algún aspecto del depósito, házmelo saber.', source_nodes=[NodeWithScore(node=TextNode(id_

Now let's use our proposal.

### **AsyncMultiQuery TSF-RAG**

For our proposal, we have implemented two systems. The first is imported under the name `AsyncMultiQueryRAG`, and the second implementation is called `AsyncMultiEmbeddingRAGV2`, which includes an `EmbeddingConfig` class to specify different embedding models.

We will show an example using the first implementation and then the second implementation. To evaluate our system, we use the second implementation.

1. **AsyncMultiQueryRAG**
2. **AsyncMultiEmbeddingRAGV2**

In [2]:
from rag import AsyncMultiQueryRAG # Old implementation
from utils import EmbeddingConfig, AsyncMultiEmbeddingRAGV2

c:\Users\gol_m\anaconda3\envs\paper_geotecnia\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


resource module not available on Windows


#### **1. AsyncMultiQuery**

In [ ]:
r = pipeline_rag.query("tipo de depósito?")

In [2]:
path_vector_storage = "./src/evaluation_retrieval/vector_storage/03_cenizas_espesado/03_cenizas_nomic"
embedding_type = "nomic"
embedding_model_name = "nomic-embed-text-v1.5"

pipeline_rag = AsyncMultiQueryRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name
)

Cargando Cross-Encoder...


In [3]:
response = pipeline_rag.query("Tipo de depósito?")
response

Desempaquetando Lista: 100%|██████████| 5/5 [00:00<?, ?it/s]


Total de nodos filtrados: 4


Response(response='El tipo de depósito mencionado en el contexto se refiere a depósitos de estéril y depósitos de relaves espesados. A continuación, se detallan las características de cada uno:\n\n1. **Depósito de Estéril (existente)**:\n   - **Estructura**: Ubicado a unos 200 metros frente a la bocamina.\n   - **Dimensión**: Ocupa una superficie aproximada de 1,5 hectáreas.\n\n2. **Depósito de Estéril (proyectado)**:\n   - **Estructura**: Nueva zona de depositación de estéril, ubicada al suroeste del depósito existente.\n   - **Capacidad**: Capacidad de depositación proyectada de 2,0 millones de toneladas.\n\n3. **Depósito de Estéril (proyecto en operación)**:\n   - **Estructura**: Construcción de un depósito para recibir el material estéril proveniente de la mina rajo.\n   - **Dimensiones**: Aproximadamente 40 hectáreas de superficie y 56 metros de altura.\n\n4. **Pilas Dump o Pilas Estáticas (proyecto pendiente)**:\n   - **Estructura**: Instalación para lixiviar mineral de baja ley 

In [5]:
from IPython.display import display, Markdown

display(Markdown(response.response))

El tipo de depósito mencionado en el contexto se refiere a depósitos de estéril y depósitos de relaves espesados. A continuación, se detallan las características de cada uno:

1. **Depósito de Estéril (existente)**:
   - **Estructura**: Ubicado a unos 200 metros frente a la bocamina.
   - **Dimensión**: Ocupa una superficie aproximada de 1,5 hectáreas.

2. **Depósito de Estéril (proyectado)**:
   - **Estructura**: Nueva zona de depositación de estéril, ubicada al suroeste del depósito existente.
   - **Capacidad**: Capacidad de depositación proyectada de 2,0 millones de toneladas.

3. **Depósito de Estéril (proyecto en operación)**:
   - **Estructura**: Construcción de un depósito para recibir el material estéril proveniente de la mina rajo.
   - **Dimensiones**: Aproximadamente 40 hectáreas de superficie y 56 metros de altura.

4. **Pilas Dump o Pilas Estáticas (proyecto pendiente)**:
   - **Estructura**: Instalación para lixiviar mineral de baja ley y ley marginal proveniente de la mina rajo sin chancar.
   - **Dimensiones**: Aproximadamente 32 hectáreas de superficie para mineral de baja ley y 40 hectáreas para mineral de ley marginal.

5. **Depósito de Relaves Espesados**:
   - **Estructura**: Se menciona un procedimiento para la operación normal del proyecto de depósito de relaves espesados, que incluye la disposición de los relaves en forma de pulpas.

Si necesitas información más específica sobre algún tipo de depósito, por favor indícalo.

What we did not mention in our paper (since it was not the focus) is that by increasing the probability of finding the pieces of text (nodes) that contain the information, we improve the response quality of the LLM.

#### **2. AsyncMultiEmbeddingRAGV2**

The benefit of the second implementation is that we can specify a list of embedding models, and the response generated by our system comes with the multiple queries generated and all the text chunks retrieved.

In [3]:
embedding_configs = [
    EmbeddingConfig(
        name="minilm",
        embedding_type="huggingface",
        model_name="sentence-transformers/all-MiniLM-L6-v2",
        path_vector_storage="./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_all_mini_L6_v2"
    ),
    EmbeddingConfig(
        name="nomic",
        embedding_type="nomic",
        model_name="nomic-embed-text-v1.5",
        path_vector_storage="./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_nomic"
    )
]


# Crear instancia del sistema RAG
pipeline_rag = AsyncMultiEmbeddingRAGV2(
    embedding_configs=embedding_configs,
    top_k=13,
    top_n=5
)

Configurando modelos de embeddings e índices...
Configurando embedding: minilm
✓ Embedding minilm configurado correctamente
Configurando embedding: nomic
✓ Embedding nomic configurado correctamente
Cargando Cross-Encoder...


In [4]:
response, metadata = pipeline_rag.query("cuál es el tipo de depósito?")

Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


Procesando embeddings:   0%|          | 0/2 [00:00<?, ?it/s]

✓ Retrieval completado para embedding: minilm

Procesando embeddings:  50%|█████     | 1/2 [00:00<00:00,  5.54it/s]

Procesando embeddings: 100%|██████████| 2/2 [00:02<00:00,  1.11s/it]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 35
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta


In [6]:
response.response

'El tipo de depósito mencionado es un "Tranque de Relaves Nº 6", que es un depósito convencional donde los relaves son procesados a través de un sistema de clasificación. En este sistema, la fracción gruesa (arena) se utiliza para construir el muro de contención del tranque, mientras que la fracción fina (lamas) se almacena en la cubeta del depósito. \n\nEl muro del tranque está compuesto principalmente de arenas, y se han realizado ensayos de penetración y triaxiales para determinar sus propiedades geotécnicas. La altura del muro se proyecta aumentar hasta 553,5 msnm, lo que permitirá depositar una cantidad adicional de relaves. \n\nEn resumen, el Tranque de Relaves Nº 6 es un depósito diseñado para almacenar relaves de la operación minera, utilizando un sistema de clasificación para separar las fracciones de material.'

In [8]:
metadata['queries_generated']

['cuál es el tipo de depósito?',
 'tipo de depósito',
 '¿Qué tipo de depósito es?',
 'dime el tipo de depósito',
 'me puedes indicar el tipo de depósito?']

In [10]:
metadata

{'queries_generated': ['cuál es el tipo de depósito?',
  'tipo de depósito',
  '¿Qué tipo de depósito es?',
  'dime el tipo de depósito',
  'me puedes indicar el tipo de depósito?'],
 'embedding_results': {'minilm': [[NodeWithScore(node=TextNode(id_='4004a208-9305-4d20-ab80-4dae0e09ad5f', embedding=None, metadata={'page_label': '7', 'file_name': 'RCA 594-Tranque 6.pdf', 'file_path': '..\\paper\\rag\\documentos\\cerro_negro\\RCA 594-Tranque 6.pdf', 'file_type': 'application/pdf', 'file_size': 163368, 'creation_date': '2025-04-08', 'last_modified_date': '2022-06-01', 'retrieval_score': np.float32(-2.7439432)}, excluded_embed_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], excluded_llm_metadata_keys=['file_name', 'file_type', 'file_size', 'creation_date', 'last_modified_date', 'last_accessed_date'], relationships={<NodeRelationship.SOURCE: '1'>: RelatedNodeInfo(node_id='d012372e-9b70-4c47-ac77-9d01a1b86b16', node_type='4'

# **Paper AsyncMultiQuery TSF-RAG**

Now we are going to replicate what is indicated in the article by evaluating each embedding model and each RAG system for the 5 TSFs.

## **5. Experimental setup and results**

### **5.2 Embeddings evaluation**

#### **1. Sand Tailing Dams (STD)**

##### **1.1 Model: All-MiniLM-L6-v2**

In [6]:
from evaluation_retrieval import EmbeddingEvaluatorV3

evaluator = EmbeddingEvaluatorV3(results_dir="./src/results/01_tranque_retrieval_evaluation")

# Una sola línea para evaluar todo
results, df = evaluator.quick_evaluate(
    model_config={"model_type": "huggingface", "model_id": "sentence-transformers/all-MiniLM-L6-v2"},
    storage_path="./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_all_mini_L6_v2",
    ground_truth_path="./src/evaluation_retrieval/ground_truth_retrieval/01_tranque_ground_truth_evaluacion.json",
    top_k=13
)

🚀 EVALUACIÓN RÁPIDA INICIADA
🤖 Cargando modelo tipo: huggingface
📁 Storage path: ./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_all_mini_L6_v2
   📦 Modelo HuggingFace: sentence-transformers/all-MiniLM-L6-v2
✅ Modelo cargado exitosamente!
📋 Cargando ground truth: ./src/evaluation_retrieval/ground_truth_retrieval/01_tranque_ground_truth_evaluacion.json
✅ Ground truth cargado exitosamente!
  📊 Tipo de depósito: tranque_de_relaves
  📈 Número de matrices: 6
  🔍 Total de parámetros: 9
    • matriz_1: 2 parámetros
    • matriz_2: 3 parámetros
    • matriz_4: 1 parámetros
    • matriz_6: 1 parámetros
    • matriz_7: 1 parámetros
    • matriz_8: 1 parámetros
🚀 Iniciando evaluación completa (top_k=13)...
📊 Total de parámetros a evaluar: 9

📈 Evaluando matriz: matriz_1 (2 parámetros)


    📊 Promedio matriz matriz_1:
       Precision: 0.3077
       Recall:    0.0889
       F1-Score:  0.1379
       AP:        0.0760

📈 Evaluando matriz: matriz_2 (3 parámetros)


    📊 Promedio matriz matriz_2:
       Precision: 0.0513
       Recall:    0.3333
       F1-Score:  0.0889
       AP:        0.1389

📈 Evaluando matriz: matriz_4 (1 parámetros)


    📊 Promedio matriz matriz_4:
       Precision: 0.1538
       Recall:    0.6667
       F1-Score:  0.2500
       AP:        0.1162

📈 Evaluando matriz: matriz_6 (1 parámetros)


    📊 Promedio matriz matriz_6:
       Precision: 0.0000
       Recall:    0.0000
       F1-Score:  0.0000
       AP:        0.0000

📈 Evaluando matriz: matriz_7 (1 parámetros)


    📊 Promedio matriz matriz_7:
       Precision: 0.0769
       Recall:    0.3333
       F1-Score:  0.1250
       AP:        0.3333

📈 Evaluando matriz: matriz_8 (1 parámetros)


    📊 Promedio matriz matriz_8:
       Precision: 0.0769
       Recall:    0.3333
       F1-Score:  0.1250
       AP:        0.0278

✅ Evaluación completa terminada!
📊 Total evaluaciones realizadas: 9

📈 RESUMEN DE EVALUACIÓN DE EMBEDDINGS
🤖 Modelo: sentence-transformers/all-MiniLM-L6-v2
🏗️  Tipo de depósito: tranque_de_relaves
📅 Fecha evaluación: 2025-10-26 22:21:52
🎯 Top-K evaluado: 13
📊 Total parámetros: 9

🌍 MÉTRICAS GLOBALES (Promedio de todos los parámetros):
--------------------------------------------------
Precision:     0.1197
Recall:        0.2790
F1-Score:      0.1158
Avg Precision: 0.1162

📊 MÉTRICAS POR MATRIZ:
--------------------------------------------------

📈 matriz_1:
  Precision:     0.3077
  Recall:        0.0889
  F1-Score:      0.1379
  Avg Precision: 0.0760

📈 matriz_2:
  Precision:     0.0513
  Recall:        0.3333
  F1-Score:      0.0889
  Avg Precision: 0.1389

📈 matriz_4:
  Precision:     0.1538
  Recall:        0.6667
  F1-Score:      0.2500
  Avg Precisi

### **5.3 Comparison of RAG systems**

In [1]:
from evaluation_rag import auto_evaluation_rag
from llms_modules import LLMEvaluator

#### **1. Naive RAG System Evaluation**

This section evaluates the 5 TSFs for the RAG system called Naive.

In [ ]:
from rag import NaiveRAG

##### **1.1 Sand Tailing Dams (STD)**

In [3]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_nomic"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/01_STD_Tranque_GT.xlsx"
nombre_excel = "01_STD_NaiveRAG_nomic.xlsx"

pipeline_rag = NaiveRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name,
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

Evaluación Matriz:   0%|          | 0/8 [00:00<?, ?it/s]

Matriz 1


Evaluación Matriz:  12%|█▎        | 1/8 [00:19<02:17, 19.71s/it]

Matriz 2


Evaluación Matriz:  25%|██▌       | 2/8 [00:28<01:19, 13.20s/it]

Matriz 3


Evaluación Matriz:  38%|███▊      | 3/8 [00:33<00:46,  9.35s/it]

Matriz 4


Evaluación Matriz:  50%|█████     | 4/8 [00:38<00:31,  7.85s/it]

Matriz 5


Evaluación Matriz:  62%|██████▎   | 5/8 [00:48<00:25,  8.47s/it]

Matriz 6


Evaluación Matriz:  75%|███████▌  | 6/8 [00:57<00:17,  8.89s/it]

Matriz 7


Evaluación Matriz:  88%|████████▊ | 7/8 [01:11<00:10, 10.53s/it]

Matriz 8


Evaluación Matriz: 100%|██████████| 8/8 [01:22<00:00, 10.37s/it]


El tiempo en evaluar fue: 83.2460584640503


##### **1.2 Filtered Tailings Deposit (FTD)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/02_kozan_filtrado/02_kozan_nomic"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/02_FTD_Filtrado_GT.xlsx"
nombre_excel = "01_FTD_NaiveRAG_nomic.xlsx"

pipeline_rag = NaiveRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name,
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

##### **1.3 Thickened Tailings Deposit (TTD)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/03_cenizas_espesado/03_cenizas_nomic"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/03_TTD_Espesado_GT.xlsx"
nombre_excel = "01_TTD_NaiveRAG_nomic.xlsx"

pipeline_rag = NaiveRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name,
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

##### **1.4 Tailings Embankments (TE)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/04_enami_embalse/04_enami_nomic"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/04_TE_Embalse_GT.xlsx"
nombre_excel = "01_TE_NaiveRAG_nomic.xlsx"

pipeline_rag = NaiveRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name,
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

##### **1.5 Paste Tailings Deposits (PTD)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/05_florida_pasta/05_florida_nomic"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/05_PTD_EnPasta_GT.xlsx"
nombre_excel = "01_PTD_NaiveRAG_nomic.xlsx"

pipeline_rag = NaiveRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name,
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

#### **2. Advanced RAG System Evaluation**

In [ ]:
from rag import AdvancedRAG

##### **2.1 Sand Tailing Dams (STD)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_nomic"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/01_STD_Tranque_GT.xlsx"
nombre_excel = "02_STD_AdvancedRAG_nomic.xlsx"

cross_encoder_model_name="cross-encoder/ms-marco-MiniLM-L-6-v2"
top_n = 5

pipeline_rag = AdvancedRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name,
    cross_encoder_model_name=cross_encoder_model_name,
    top_n=top_n
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

Evaluación Matriz:   0%|          | 0/8 [00:00<?, ?it/s]

Matriz 1


Evaluación Matriz:  12%|█▎        | 1/8 [00:23<02:43, 23.29s/it]

Matriz 2


Evaluación Matriz:  25%|██▌       | 2/8 [00:37<01:47, 17.91s/it]

Matriz 3


Evaluación Matriz:  38%|███▊      | 3/8 [00:43<01:01, 12.36s/it]

Matriz 4


Evaluación Matriz:  50%|█████     | 4/8 [00:50<00:41, 10.45s/it]

Matriz 5


Evaluación Matriz:  62%|██████▎   | 5/8 [01:01<00:31, 10.51s/it]

Matriz 6


Evaluación Matriz:  75%|███████▌  | 6/8 [01:12<00:21, 10.90s/it]

Matriz 7


Evaluación Matriz:  88%|████████▊ | 7/8 [01:29<00:12, 12.66s/it]

Matriz 8


Evaluación Matriz: 100%|██████████| 8/8 [01:45<00:00, 13.14s/it]

El tiempo en evaluar fue: 105.29558682441711


##### **2.2 Filtered Tailings Deposit (FTD)**

In [ ]:
path_vector_storage="./src/evaluation_retrieval/vector_storage/02_kozan_filtrado/02_kozan_nomic"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/02_FTD_Filtrado_GT.xlsx"
nombre_excel = "02_FTD_NaiveRAG_nomic.xlsx"

cross_encoder_model_name="cross-encoder/ms-marco-MiniLM-L-6-v2"
top_n = 5

pipeline_rag = AdvancedRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name,
    cross_encoder_model_name=cross_encoder_model_name,
    top_n=top_n
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

##### **2.3 Thickened Tailings Deposit (TTD)**

In [ ]:
path_vector_storage="./src/evaluation_retrieval/vector_storage/03_cenizas_espesado/03_cenizas_nomic"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/03_TTD_Espesado_GT.xlsx"
nombre_excel = "02_TTD_AdvancedRAG_nomic.xlsx"

cross_encoder_model_name="cross-encoder/ms-marco-MiniLM-L-6-v2"
top_n = 5

pipeline_rag = AdvancedRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name,
    cross_encoder_model_name=cross_encoder_model_name,
    top_n=top_n
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

##### **2.4 Tailings Embankments (TE)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/04_enami_embalse/04_enami_nomic"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/04_TE_Embalse_GT.xlsx"
nombre_excel = "02_TE_AdvancedRAG_nomic.xlsx"

cross_encoder_model_name="cross-encoder/ms-marco-MiniLM-L-6-v2"
top_n = 5

pipeline_rag = AdvancedRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name,
    cross_encoder_model_name=cross_encoder_model_name,
    top_n=top_n
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

##### **2.5 Paste Tailings Deposits (PTD)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/05_florida_pasta/05_florida_nomic"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/05_PTD_EnPasta_GT.xlsx"
nombre_excel = "02_PTD_AdvancedRAG_nomic.xlsx"

cross_encoder_model_name="cross-encoder/ms-marco-MiniLM-L-6-v2"
top_n = 5

pipeline_rag = AdvancedRAG(
    path_vector_storage=path_vector_storage,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name,
    cross_encoder_model_name=cross_encoder_model_name,
    top_n=top_n
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

#### **3. Hybrid RAG System Evaluation**

In [ ]:
from rag import HybridRAG

##### **3.1 Sand Tailing Dams (STD)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_nomic"
path_bm25 = "./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_bm25"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/01_STD_Tranque_GT.xlsx"
nombre_excel = "03_STD_HybridRAG_nomic.xlsx"

pipeline_rag = HybridRAG(
    path_vector_storage=path_vector_storage,
    path_bm25=path_bm25,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

Evaluación Matriz:   0%|          | 0/8 [00:00<?, ?it/s]

Matriz 1


Evaluación Matriz:  12%|█▎        | 1/8 [00:18<02:09, 18.49s/it]

Matriz 2


Evaluación Matriz:  25%|██▌       | 2/8 [00:29<01:23, 13.98s/it]

Matriz 3


Evaluación Matriz:  38%|███▊      | 3/8 [00:32<00:46,  9.25s/it]

Matriz 4


Evaluación Matriz:  50%|█████     | 4/8 [00:39<00:31,  8.00s/it]

Matriz 5


Evaluación Matriz:  62%|██████▎   | 5/8 [00:50<00:27,  9.13s/it]

Matriz 6


Evaluación Matriz:  75%|███████▌  | 6/8 [00:57<00:16,  8.38s/it]

Matriz 7


Evaluación Matriz:  88%|████████▊ | 7/8 [01:13<00:11, 11.12s/it]

Matriz 8


Evaluación Matriz: 100%|██████████| 8/8 [01:22<00:00, 10.28s/it]

El tiempo en evaluar fue: 82.45992946624756


##### **3.2 Filtered Tailings Deposit (FTD)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/02_kozan_filtrado/02_kozan_nomic"
path_bm25 = "./src/evaluation_retrieval/vector_storage/02_kozan_filtrado/02_kozan_filtrado_bm25"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/02_FTD_Filtrado_GT.xlsx"
nombre_excel = "03_FTD_HybridRAG_nomic.xlsx"


pipeline_rag = HybridRAG(
    path_vector_storage=path_vector_storage,
    path_bm25=path_bm25,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

##### **3.3 Thickened Tailings Deposit (TTD)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/03_cenizas_espesado/03_cenizas_nomic"
path_bm25 = "./src/evaluation_retrieval/vector_storage/03_cenizas_espesado/03_cenizas_espesado_bm25"

embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/03_TTD_Espesado_GT.xlsx"
nombre_excel = "03_TTD_HybridRAG_nomic.xlsx"


pipeline_rag = HybridRAG(
    path_vector_storage=path_vector_storage,
    path_bm25=path_bm25,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

##### **3.4 Tailings Embankments (TE)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/04_enami_embalse/04_enami_nomic"
path_bm25 = "./src/evaluation_retrieval/vector_storage/04_enami_embalse/04_enami_embalse_bm25"

embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/04_TE_Embalse_GT.xlsx"
nombre_excel = "03_TE_HybridRAG_nomic.xlsx"


pipeline_rag = HybridRAG(
    path_vector_storage=path_vector_storage,
    path_bm25=path_bm25,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

##### **3.5 Paste Tailings Deposits (PTD)**

In [ ]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/05_florida_pasta/05_florida_nomic"
path_bm25 = "./src/evaluation_retrieval/vector_storage/05_florida_pasta/05_florida_pasta_bm25"

embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/05_PTD_EnPasta_GT.xlsx"
nombre_excel = "03_PTD_HybridRAG_nomic.xlsx"


pipeline_rag = HybridRAG(
    path_vector_storage=path_vector_storage,
    path_bm25=path_bm25,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

#### **4. Advanced-Hybrid RAG System Evaluation**

In [ ]:
from rag import AdvancedHybridRAG

##### **4.1 Sand Tailing Dams (STD)**

In [7]:
import time 

path_vector_storage="./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_nomic"
path_bm25 = "./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_bm25"
embedding_type="nomic"
embedding_model_name="nomic-embed-text-v1.5"

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/01_STD_Tranque_GT.xlsx"
nombre_excel = "04_STD_AdvancedHybridRAG_nomic.xlsx"

cross_encoder_model_name="cross-encoder/ms-marco-MiniLM-L-6-v2"
top_n = 5

pipeline_rag = AdvancedHybridRAG(
    path_vector_storage=path_vector_storage,
    path_bm25=path_bm25,
    embedding_type=embedding_type,
    embedding_model_name=embedding_model_name,
    cross_encoder_model_name=cross_encoder_model_name,
    top_n=top_n
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

Evaluación Matriz:   0%|          | 0/8 [00:00<?, ?it/s]

Matriz 1


Evaluación Matriz:  12%|█▎        | 1/8 [00:21<02:29, 21.39s/it]

Matriz 2


Evaluación Matriz:  25%|██▌       | 2/8 [00:35<01:42, 17.03s/it]

Matriz 3


Evaluación Matriz:  38%|███▊      | 3/8 [00:40<00:57, 11.60s/it]

Matriz 4


Evaluación Matriz:  50%|█████     | 4/8 [00:47<00:39,  9.97s/it]

Matriz 5


Evaluación Matriz:  62%|██████▎   | 5/8 [00:57<00:29,  9.92s/it]

Matriz 6


Evaluación Matriz:  75%|███████▌  | 6/8 [01:06<00:19,  9.67s/it]

Matriz 7


Evaluación Matriz:  88%|████████▊ | 7/8 [01:22<00:11, 11.62s/it]

Matriz 8


Evaluación Matriz: 100%|██████████| 8/8 [01:31<00:00, 11.48s/it]


El tiempo en evaluar fue: 92.1056878566742


##### **4.2 Filtered Tailings Deposit (FTD)**

##### **4.3 Thickened Tailings Deposit (TTD)**

##### **4.4 Tailings Embankments (TE)**

##### **4.5 Paste Tailings Deposits (PTD)**

#### **5. AsyncMultiQuery RAG System Evaluation**

In [2]:
from utils import EmbeddingConfig, AsyncMultiEmbeddingRAGV2

resource module not available on Windows


c:\Users\gol_m\anaconda3\envs\paper_geotecnia\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


##### **5.1 Sand Tailing Dams (STD)**

In [5]:
import time 

embedding_configs = [
    EmbeddingConfig(
        name="minilm",
        embedding_type="huggingface",
        model_name="sentence-transformers/all-MiniLM-L6-v2",
        path_vector_storage="./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_all_mini_L6_v2"
    ),
    EmbeddingConfig(
        name="nomic",
        embedding_type="nomic",
        model_name="nomic-embed-text-v1.5",
        path_vector_storage="./src/evaluation_retrieval/vector_storage/01_cerro_negro_tranque/01_cerro_negro_nomic"
    )
]

path_ground_truth_rag = "./src/evaluation_rag/rag_ground_truth/01_STD_Tranque_GT.xlsx"
nombre_excel = "05_STD_AsyncMultiQuery_nomic.xlsx"

# Crear instancia del sistema RAG
pipeline_rag = AsyncMultiEmbeddingRAGV2(
    embedding_configs=embedding_configs,
    top_k=13,
    top_n=5
)

llm_evaluador = LLMEvaluator()

inicio = time.time()
auto_evaluation_rag(
    path_ground_truth=path_ground_truth_rag,
    name_excel=nombre_excel,
    pipeline_rag=pipeline_rag,
    llm_evaluador=llm_evaluador
)
fin = time.time()
tiempo = fin - inicio
print("El tiempo en evaluar fue:", tiempo)

Configurando modelos de embeddings e índices...
Configurando embedding: minilm
✓ Embedding minilm configurado correctamente
Configurando embedding: nomic
✓ Embedding nomic configurado correctamente
Cargando Cross-Encoder...


Evaluación Matriz:   0%|          | 0/8 [00:00<?, ?it/s]

Matriz 1
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:02<00:00,  1.09s/it]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 30
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  2.33it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 30
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta


Evaluación Matriz:  12%|█▎        | 1/8 [00:45<05:19, 45.66s/it]

Matriz 2
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  3.71it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 26
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  2.41it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 28
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  3.17it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 40
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta


Evaluación Matriz:  25%|██▌       | 2/8 [01:37<04:56, 49.38s/it]

Matriz 3
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  2.38it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 39
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta


Evaluación Matriz:  38%|███▊      | 3/8 [01:58<03:02, 36.41s/it]

Matriz 4
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  2.33it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 24
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta


Evaluación Matriz:  50%|█████     | 4/8 [02:20<02:03, 30.77s/it]

Matriz 5
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  3.23it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 28
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  3.79it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 35
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta


Evaluación Matriz:  62%|██████▎   | 5/8 [03:07<01:49, 36.55s/it]

Matriz 6
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  3.97it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 37
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta


Evaluación Matriz:  75%|███████▌  | 6/8 [03:30<01:03, 31.79s/it]

Matriz 7
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  3.38it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 32
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  3.09it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 52
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta


Evaluación Matriz:  88%|████████▊ | 7/8 [04:32<00:41, 41.86s/it]

Matriz 8
Etapa 1 - Total de querys generadas: 5
Etapa 2 - Retrieval con múltiples embeddings
Iniciando retrieval con múltiples embeddings...


✓ Retrieval completado para embedding: minilm


Procesando embeddings: 100%|██████████| 2/2 [00:00<00:00,  3.53it/s]

✓ Retrieval completado para embedding: nomic
Etapa 3 - Agregación y filtrado de nodos
Agregando y filtrando nodos de todos los embeddings...
Procesando nodos de embedding: minilm
Procesando nodos de embedding: nomic
Total de nodos antes del filtrado: 130
Total de nodos después del filtrado: 49
Etapa 4 - Reranking


Etapa 5 - Generación de respuesta


Evaluación Matriz: 100%|██████████| 8/8 [05:04<00:00, 38.08s/it]

El tiempo en evaluar fue: 304.95287132263184


##### **5.2 Filtered Tailings Deposit (FTD)**

##### **5.3 Thickened Tailings Deposit (TTD)**

##### **5.4 Tailings Embankments (TE)**

##### **5.5 Paste Tailings Deposits (PTD)**